Plugins/BundleEditorPlugin/BundleEditor.cs

```diff
@@ -1,249 +1,578 @@
+
+        public BundleEditor()
+        {
+            foreach (var type in Assembly.GetCallingAssembly().GetTypes())
+            {
+                if (type.IsSubclassOf(typeof(AddToBundleExtension)))
+                {
+                    var extension = (AddToBundleExtension)Activator.CreateInstance(type);
+                    addToBundleExtensions.Add(extension.AssetType, extension);
+                }
+                else if (type.IsSubclassOf(typeof(RemoveFromBundleExtension)))
+                {
+                    var extension = (RemoveFromBundleExtension)Activator.CreateInstance(type);
+                    removeFromBundleExtensions.Add(extension.AssetType, extension);
+                }
+            }
+            addToBundleExtensions.Add("null", new AddToBundleExtension());
+            removeFromBundleExtensions.Add("null", new RemoveFromBundleExtension());
+
+            AddToBundleCommand = new RelayCommand(
+                (o) =>
+                {
+                    EbxAssetEntry entry = App.EditorWindow.DataExplorer.SelectedAsset as EbxAssetEntry;
+                    BundleEntry bentry = bundlesListBox.SelectedItem as BundleEntry;
+
+                    if (!entry.Bundles.Contains(App.AssetManager.GetBundleId(bentry)) && !entry.AddedBundles.Contains(App.AssetManager.GetBundleId(bentry)))
+                    {
+                        string key = entry.Type;
+                        if (!addToBundleExtensions.ContainsKey(entry.Type))
+                        {
+                            key = "null";
+                            foreach (string typekey in addToBundleExtensions.Keys)
+                            {
+                                if (TypeLibrary.IsSubClassOf(entry.Type, typekey))
+                                {
+                                    key = typekey;
+                                    break;
+                                }
+                            }
+                        }
+                        addToBundleExtensions[key].AddToBundle(entry, bentry);
+                    }
+
+                    else
+                    {
+                        App.Logger.LogError("Asset is already in {0}", bentry.Name);
+                    }
+
+                    RefreshExplorer();
+                    App.EditorWindow.DataExplorer.RefreshItems();
+
+                    dataExplorer.SelectAsset(entry);
+                },
+                (o) =>
+                {
+                    return App.EditorWindow.DataExplorer.SelectedAsset != null && bundlesListBox.SelectedItem != null;
+                });
+
+            RemoveFromBundleCommand = new RelayCommand(
+                (o) =>
+                {
+                    EbxAssetEntry entry = App.EditorWindow.DataExplorer.SelectedAsset as EbxAssetEntry;
+                    BundleEntry bentry = bundlesListBox.SelectedItem as BundleEntry;
+
+                    if (!entry.Bundles.Contains(App.AssetManager.GetBundleId(bentry)))

```

Comment:
shouldnt u check if the added bundles contains the bundle?

Reply:
ah yeah that would be better, I was just specifically thinking of the scenario where the user tries to remove an existing one, not anything else